stl/inc/format

```diff
@@ -163,6 +163,697 @@ concept _Parse_spec_callbacks = _Parse_align_callbacks<_Ty, _CharT>
+private:
+    const _CharT* _First = nullptr;
+    const _CharT* _Last  = nullptr;
+    const _CharT* _Next  = nullptr;
+    char32_t _Val        = 0;
+
+public:
+    using iterator_concept = forward_iterator_tag;
+    using value_type       = char32_t;
+    using difference_type  = ptrdiff_t;
+
+    constexpr _Unicode_codepoint_iterator(const _CharT* _First_val, const _CharT* _Last_val)
+        : _First(_First_val), _Last(_Last_val) {
+        _Next = _Decode_utf(_First, _Last, _Val);
+    }
+
+    constexpr _Unicode_codepoint_iterator(const basic_string_view<_CharT>& _Str)
+        : _Unicode_codepoint_iterator(_Str.data(), _Str.data() + _Str.size()) {}
+
+    constexpr _Unicode_codepoint_iterator() = default;
+
+    constexpr _Unicode_codepoint_iterator& operator++() noexcept {
+        _First = _Next;
+        if (_First != _Last) {
+            _Next = _STD _Decode_utf(_First, _Last, _Val);
+        }
+
+        return *(this);
+    }
+    constexpr _Unicode_codepoint_iterator operator++(int) noexcept {
+        auto _Old = *this;
+        ++*this;
+        return _Old;
+    }
+    constexpr value_type operator*() const noexcept {
+        return _Val;
+    }
+
+    constexpr const _CharT* _Position() const {
+        return _First;
+    }
+
+    constexpr bool operator==(_Unicode_codepoint_end_iterator) const noexcept {
+        if (_First == _Last) {
+            return true;
+        }
+        return false;
+    }
+    constexpr bool operator==(const _Unicode_codepoint_iterator& _Other) const {
+        return (_First == _Other._First) && (_Last == _Other._Last);
+    }
+};
+
+
+// implements a DFA matching the regex on the left side of rule GB11. The DFA is:
+//
+// +---+   ExtPic      +---+    ZWJ        +---+
+// | 1 +---------------> 2 +---------------> 3 |
+// +---+               ++-^+               +---+
+//                      | |
+//                      +-+
+//                      Extend
+//
+// Note state 3 is never explicitly transitioned to, since it's the "accept" state, we just
+// transition back to state 1 and return true.
+class _GB11_LeftHand_regex {
+    enum { _Start, _ExtPic } _State = _Start;
+
+public:
+    constexpr bool _Match(
+        const _Grapheme_Break_Property_Values _Left_gbp, _Extended_Pictographic_Property_Values _Left_ExtPic) noexcept {
+        switch (_State) {
+        case _Start:
+            if (_Left_ExtPic == _Extended_Pictographic_Property_Values::_Extended_Pictographic_value) {
+                _State = _ExtPic;
+            }
+            return false;
+        case _ExtPic:
+            if (_Left_gbp == _Grapheme_Break_Property_Values::_ZWJ_value) {
+                _State = _Start;
+                return true;
+            } else if (_Left_gbp != _Grapheme_Break_Property_Values::_Extend_value) {
+                _State = _Start;
+                return false;
+            }
+            return false;
+        default:
+            _STL_ASSERT(false, "Invalid GB11 state.");

```

Comment:
Should this be an `_STL_ASSERT` (which detects users with precondition violations) or an `_STL_INTERNAL_CHECK` (which detects our logic errors, and should never be triggered by arbitrary well-formed user input)?